### PR TITLE
ASSETS-88901 : (Backend) Campaign Page: Filters

### PR DESCRIPTION
**Pull request by TyroneAEM**
Workfront Task :  https://experience.adobe.com/#/@wfadoberm/so:adoberm-Production/workfront/team/64120499002edcb6e4ef1dbc3e644241/iterations

Test URLs: https://assets-88901--adobe-gmo--hlxsites.hlx.page/drafts/tyrone/campaigns#
Screenshot 2024-04-30 at 4 29 48 PM
Screenshot 2024-04-30 at 4 29 59 PM

AEM GraphQL Persisted Queries Created/Used

Used for Campaign List
https://author-p108396-e1046543.adobeaemcloud.com/graphql/execute.json/gmo/getAllCampaigns;first=;cursor=;filter=

Used for Campaign total records count
https://author-p108396-e1046543.adobeaemcloud.com/graphql/execute.json/gmo/getCampaignNameFilter;filter=

Used for Campaign Name search auto complete list
https://author-p108396-e1046543.adobeaemcloud.com/graphql/execute.json/gmo/getCampaignNames;campaignName=

### CODE REVIEW

**Review by mdickson-adbe — APPROVED** [review]

File: blocks/gmo-campaign-list/gmo-campaign-list.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { readBlockConfig } from '../../scripts/lib-franklin.js';
 import { decorateIcons } from '../../scripts/lib-franklin.js';
-import { graphqlAllCampaigns, graphqlCampaignCount } from '../../scripts/graphql.js';
+import { graphqlAllCampaignsFilter, graphqlAllCampaigns, graphqlCampaignCount, generateFilterJSON } from '../../scripts/graphql.js';
 
 const icon = 'https://delivery-p108396-e1046543.adobeaemcloud.com/adobe/assets/deliver/urn:aaid:aem:acdaa42f-00ae-42f4-97e5-8309c42d9076/marketing-hub-102023-lockup-video.png'
```
Comment (shiv-gup) [inline blocks/gmo-campaign-list/gmo-campaign-list.js:5]: @TyroneAEM Are we using this icon?

File: scripts/graphql.js [hunk 2]
```diff
@@ -1,13 +1,72 @@
...
+      }
+      return response.json();
+  }).then(data => {
+      return data; // Make sure to return the data so that the promise resolves with it
+  }).catch(error => {
+      console.error('Error fetching data: ', error);
+      throw error; // Rethrow or handle error as appropriate
+  });
+}
+export async function graphqlStatusList() {
```
Comment (shiv-gup) [inline scripts/graphql.js:32]: `graphqlProductList` and `graphqlStatusList` function definition looks same except the query name. Can you combine these methods and have query name passed as parameter?

File: scripts/graphql.js [hunk 3]
```diff
@@ -1,13 +1,72 @@
...
+
+  const baseApiUrl = `${await getGraphqlEndpoint()}/graphql/execute.json`;
```
Comment (shiv-gup) [inline scripts/graphql.js:34]: let's create a class level variable for `baseApiUrl` like how we have `deliveryEndpointURL` in `polaris.js`

File: scripts/graphql.js [hunk 4]
```diff
@@ -139,3 +239,70 @@ async function getGraphqlEndpoint() {
...
+ *			{
+ *				"value": "<Value>",
+ *			}
+ *		]
+ *  @returns {string} GraphQL Filter JSON Object.
+ */
+
+export function generateFilterJSON(filterParams) {
+
+  /* Example of valid array of parameters
```
Comment (shiv-gup) [inline scripts/graphql.js:272]: can you remove this comment as this is already added in line number 250

File: blocks/gmo-campaign-list/gmo-campaign-list.js (hunk 1 above)
Comment (TyroneAEM) [reply blocks/gmo-campaign-list/gmo-campaign-list.js:5]: No we are not, I'll delete it

File: scripts/graphql.js (hunk 3 above)
Comment (TyroneAEM) [reply scripts/graphql.js:34]: created class level variable for baseApiUrl and projectId

**Review by shiv-gup — APPROVED** [review]

**Pull request merged**